lib/training/training_resource_query_object.rb

```diff
@@ -0,0 +1,73 @@
+# frozen_string_literal: true
+
+require_dependency "#{Rails.root}/lib/utils/string_utils"
+
+class TrainingResourceQueryObject
+  EXCERPT_LENGTH = 60
+
+  def initialize(current_user, search = nil)
+    @current_user = current_user
+    @search = search
+  end
+
+  def all_libraries
+    libraries = TrainingLibrary.all.sort_by do |library|
+      library.slug == focused_library_slug ? 0 : 1
+    end
+
+    [focused_library_slug, libraries]
+  end
+
+  def selected_slides_and_excerpt
+    slugs = search_content_in_slides.map do |slide|
+      { slide: slide.slug, title: slide.title, excerpt: excerpt(slide.content) }
+    end
+    add_libr_and_mod_slugs(slugs)
+
+    slide_paths_with_excerpt(slugs)
+  end
+
+  private
+
+  def add_libr_and_mod_slugs(slugs)
+    slugs.each do |slug|
+      slug[:module], slug[:module_name] =
+        module_of_a_slide(slug[:slide]).then { |tm| [tm.slug, tm.name] }
+      slug[:library] = library_of_a_module(slug[:module]).slug
```

Comment:
I'm getting errors in some cases from this, when no library is found.

Reply:
Aaaah, I was not aware of that. I thought each module belonged to a library. Thank you for the clarification.
Just to be sure: do I have to exclude those slides(belonging to a module without library) from the results ?

Reply:
Hmm... I think including them in the results would be fine (the links just need to be munged to avoid broken paths).

Reply:
I have added some "sanitazation" (?). I realised that the search was made on text inside tags like img and href and afterwards were displayed (eg search on 'wikipedia' displayed results on url containing wikipedia ). Also I strip markdown link too.